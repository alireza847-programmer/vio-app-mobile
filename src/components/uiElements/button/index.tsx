import React from 'react';
import Button from './style';
import {VButtonProps} from 'types/components/uiElements/button';
import VText from '../text';
import {theme} from 'themes/emotion';
import VRow from '../row';

const VButton = (props: VButtonProps) => {
  const {
    mode = 'primary',
    subTitle,
    title,
    styled = 'FULL_WIDTH',
    icon,
    rightIcon,
    onPress,
    testID,
    disabled = false,
    width,
  } = props;

  const renderContentColor = () => {
    if (disabled) {
      return theme.colors.textColor.button.disabled;
    }
    if (styled === 'TEXT') {
      return theme.colors.button[mode];
    }
    return theme.colors.textColor.button[mode];
  };

  return (
    <Button
      width={width}
      disabled={disabled}
      testID={testID}
      activeOpacity={0.8}
      onPress={onPress}
      styled={styled}
      mode={mode}>
      {icon && (
        <VRow {...(title && {marginRighRatio: 2})} fullWidth={false}>
          {icon(renderContentColor())}
        </VRow>
      )}
      <VText
        typography="semiBold16"
        color={renderContentColor()}
        marginTopRatio={0}>
        {title}
      </VText>
      {subTitle && (
        <VText
          typography="regular16"
          color={renderContentColor()}
          marginLeftRatio={1.2}
          marginTopRatio={0}>
          {subTitle}
        </VText>
      )}
      {rightIcon && (
        <VRow {...(title && {marginLeftRatio: 2})} fullWidth={false}>
          {rightIcon(renderContentColor())}
        </VRow>
      )}
    </Button>
  );
};
export default VButton;
