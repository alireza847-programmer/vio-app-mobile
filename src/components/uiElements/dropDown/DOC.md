# VDropDown Component Documentation

## Overview

The `VDropDown` component is a versatile dropdown selector designed for React Native applications. It provides an intuitive way to present a list of options to the user and capture their selection.

## Usage

```jsx
import React from 'react';
import VDropDown from './path-to-your-component/VDropDown'; // Adjust the import path based on your project structure

const MyComponent = () => {
  const data = ['Option 1', 'Option 2', 'Option 3'];
  const [selectedValue, setSelectedValue] = React.useState(''); // Track the selected value in your component's state

  const handleChange = value => {
    setSelectedValue(value);
    // Handle any additional logic here
  };

  return (
    <VDropDown
      data={data}
      placeHolder="Select an option"
      onChange={handleChange}
      defaultValue={selectedValue} // Set this to preselect a value
    />
  );
};
```

## Props

### VDropDown Props

- `data` (Array): An array of strings representing the options available in the dropdown.
- `placeHolder` (string): The placeholder text to display when no option is selected.
- `onChange` (function): A callback function triggered when the user selects an option. It receives the selected value as an argument.
- `defaultValue` (string): The default value to preselect in the dropdown.

## Example

```jsx
import React, {useState} from 'react';
import {View} from 'react-native';
import VDropDown from './path-to-your-component/VDropDown'; // Adjust the import path based on your project structure

const ExampleScreen = () => {
  const data = ['Option 1', 'Option 2', 'Option 3'];
  const [selectedValue, setSelectedValue] = useState('');

  const handleChange = value => {
    setSelectedValue(value);
  };

  return (
    <View>
      <VDropDown
        data={data}
        placeHolder="Select an option"
        onChange={handleChange}
        defaultValue={selectedValue}
      />
    </View>
  );
};

export default ExampleScreen;
```

## Style

The styling for the `VDropDown` component is managed internally and follows a clean and consistent design to ensure compatibility with various themes and styles.

## Modal

The component uses an internal modal to display the options. It renders a `WheelPicker` for easy selection and provides a confirm button to finalize the choice. The modal is automatically displayed when the user taps the dropdown.
