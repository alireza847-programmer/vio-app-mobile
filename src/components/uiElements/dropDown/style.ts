import styled from '@emotion/native';
import {Dropdown as EDropDown} from 'react-native-element-dropdown';

export const DropDown = styled(EDropDown)<{isFocous: boolean}>(
  ({theme, isFocous}) => ({
    borderColor: isFocous
      ? theme.colors.button.primary
      : theme.colors.stroke.input,
    borderWidth: 1,
    borderRadius: theme.radius * 2,
    paddingHorizontal: theme.HorizontalRem * 4,
    paddingVertical: theme.verticalRem * 2,
  }),
);
