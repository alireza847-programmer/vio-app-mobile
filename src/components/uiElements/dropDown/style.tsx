import styled from '@emotion/native';
import VRow from '../row';

export const Container = styled(VRow)(({theme}) => ({
  borderWidth: 1,
  borderColor: theme.colors.stroke.input,
  paddingLeft: theme.HorizontalRem * 4,
  paddingRight: theme.HorizontalRem * 2,
  paddingVertical: theme.verticalRem * 2,
  borderRadius: theme.radius * 2,
}));
