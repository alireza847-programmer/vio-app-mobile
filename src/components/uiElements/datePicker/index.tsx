import {theme} from '@storybook/react-native';
import React from 'react';
import {View} from 'react-native';
import Calendar from 'react-native-calendar-range-picker';
import {VDatePickerProps} from 'types/components/uiElements/datePicker';

const VDatePicker = (props: VDatePickerProps) => {
  const {...rest} = props;
  return (
    <View style={{width: '100%'}}>
      <Calendar disabledBeforeToday pastYearRange={0} {...rest} />
    </View>
  );
};

export default VDatePicker;
