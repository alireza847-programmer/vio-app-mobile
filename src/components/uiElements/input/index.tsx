import React from 'react';
import Input from './style';
import {VInputProps} from 'types/components/uiElements/input';
import {theme} from 'themes/emotion';

const VInput = (props: VInputProps) => {
  const {...rest} = props;
  return (
    <Input
      placeholderTextColor={theme.colors.textColor.placeHolder}
      {...rest}
    />
  );
};

export default VInput;
