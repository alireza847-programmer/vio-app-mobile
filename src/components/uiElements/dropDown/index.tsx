import React, {Fragment, useEffect, useState} from 'react';
import {VDropDownProps} from 'types/components/uiElements/dropDown';
import {theme} from 'themes/emotion';
import {ChevronSvg} from 'assets/svgs';
import {Container} from './style';
import WheelPicker from 'react-native-wheely';
import VModal from '../modal';
import VText from '../text';
import VRow from '../row';
import VButton from '../button';
import {TouchableOpacity} from 'react-native';

const VDropDown = (props: VDropDownProps) => {
  const {data, placeHolder, onChange, defaultValue} = props;
  const [visibleModal, setVisibleModal] = useState(false);
  const [value, setValue] = useState('');
  const [selectedIndex, setSelectedIndex] = useState(0);
  useEffect(() => {
    if (defaultValue) {
      setValue(String(defaultValue));
      setSelectedIndex(data.indexOf(String(defaultValue)));
    } else {
      setValue(String(placeHolder));
    }
  }, [placeHolder, defaultValue]);
  const onConfirm = () => {
    setVisibleModal(false);
    setValue(data[selectedIndex]);
    onChange(data[selectedIndex]);
  };
  const renderModal = () => {
    return (
      <VModal
        withClose
        withHeader
        title="Child 1 age"
        onClose={() => setVisibleModal(false)}
        isVisible={visibleModal}>
        <WheelPicker
          selectedIndex={selectedIndex}
          options={data}
          onChange={index => setSelectedIndex(index)}
          containerStyle={{width: '90%'}}
          opacityFunction={x => 0.3}
          selectedIndicatorStyle={{
            backgroundColor: theme.colors.button.secondary,
            borderTopWidth: 1,
            borderBottomWidth: 1,
            borderColor: theme.colors.button.primary,
          }}
          visibleRest={1}
          itemStyle={{
            backgroundColor: 'transparent',
          }}
          itemTextStyle={{
            ...theme.fonts.semiBold18,
          }}
        />
        <VRow paddingVerticalRatio={4} justifyContent="center">
          <VButton onPress={onConfirm} width={'90%'} title="CONFIRM" />
        </VRow>
      </VModal>
    );
  };
  return (
    <Fragment>
      <TouchableOpacity
        activeOpacity={0.8}
        onPress={() => setVisibleModal(true)}>
        <Container
          alignItems="center"
          justifyContent="space-between"
          fullWidth={false}>
          <VText
            width={theme.HorizontalRem * 10}
            textAlign="left"
            typography="semiBold16">
            {value}
          </VText>
          <VRow fullWidth={false} marginLeftRatio={3}>
            <ChevronSvg fill={theme.colors.button.primary} />
          </VRow>
        </Container>
      </TouchableOpacity>
      {renderModal()}
    </Fragment>
  );
};
export default VDropDown;
