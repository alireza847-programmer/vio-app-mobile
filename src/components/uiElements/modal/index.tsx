import Modal from 'react-native-modal';
import React, {PropsWithChildren} from 'react';
import {VModalProps} from 'types/components/uiElements/modal';
import {HeaderContainer, Wrapper, modalStyle} from './style';
import {theme} from 'themes/emotion';
import {CloseSvg} from 'assets/svgs';
import VText from '../text';
import VButton from '../button';

const VModal = (props: PropsWithChildren<VModalProps>) => {
  const {children, isVisible, onClose, title, withClose, withHeader} = props;
  return (
    <Modal
      animationIn="slideInUp"
      animationOut={'slideOutDown'}
      style={modalStyle}
      testID="modal"
      hasBackdrop={true}
      onBackButtonPress={onClose}
      onBackdropPress={() => onClose()}
      isVisible={isVisible}>
      <Wrapper>
        {withHeader && (
          <HeaderContainer
            justifyContent="space-between"
            paddingHorizontalRatio={4}
            paddingVerticalRatio={6}>
            {withClose && (
              <VButton
                onPress={onClose}
                testID="close-button"
                styled="TEXT"
                icon={fill => <CloseSvg fill={theme.colors.button.primary} />}
              />
            )}
            <VText typography="semiBold16">{title}</VText>
            <CloseSvg fill={'transparent'} />
          </HeaderContainer>
        )}
        {children}
      </Wrapper>
    </Modal>
  );
};
export default VModal;
