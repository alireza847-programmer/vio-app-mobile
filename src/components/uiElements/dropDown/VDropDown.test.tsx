import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import VDropDown from './index';
import {it, describe, expect, jest} from '@jest/globals';
import {ThemeProvider} from '@emotion/react';
import {theme} from 'themes/emotion';
import {VDropDownProps} from 'types/components/uiElements/dropDown';

const Component = (props: VDropDownProps) => (
  <ThemeProvider theme={theme}>
    <VDropDown {...props} />
  </ThemeProvider>
);

describe('VDropDown component', () => {
  const data = ['Option 1', 'Option 2', 'Option 3'];
  const onChangeMock = jest.fn();

  it('matches snapshot with default props', () => {
    const {toJSON} = render(
      <Component
        title="Select an option"
        data={data}
        placeHolder="Select an option"
        onChange={onChangeMock}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('opens modal when pressed', () => {
    const {getByTestId} = render(
      <Component
        title="Select an option"
        data={data}
        placeHolder="Select an option"
        onChange={onChangeMock}
      />,
    );

    const touchable = getByTestId('dropdown-touchable');
    fireEvent.press(touchable);

    const modal = getByTestId('modal-backdrop');
    expect(modal).toBeDefined();
  });

  it('selects a value from the modal', () => {
    const {getByTestId, getByText} = render(
      <Component
        title="Select an option"
        data={data}
        placeHolder="Select an option"
        onChange={onChangeMock}
      />,
    );

    const touchable = getByTestId('dropdown-touchable');
    fireEvent.press(touchable);

    const optionToSelect = getByText(data[1]);
    fireEvent.press(optionToSelect);

    // Ensure that the selected value is displayed
    expect(getByText(data[1])).toBeDefined();
  });

  it('calls onChange when confirming the selection', () => {
    const {getByTestId} = render(
      <Component
        title="Select an option"
        data={data}
        placeHolder="Select an option"
        onChange={onChangeMock}
      />,
    );

    const touchable = getByTestId('dropdown-touchable');
    fireEvent.press(touchable);

    const confirmButton = getByTestId('confirm-button');
    fireEvent.press(confirmButton);

    // Ensure that onChange is called
    expect(onChangeMock).toHaveBeenCalled();
  });
});
