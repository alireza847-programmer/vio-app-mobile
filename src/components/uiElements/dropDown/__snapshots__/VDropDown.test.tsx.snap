// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VDropDown component matches snapshot with custom data and placeholder 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      {
        "zIndex": 5000,
      },
      {
        "width": "40%",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onLayout={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#BED2E9",
        "borderRadius": 15,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "minHeight": 50,
        "opacity": 1,
        "paddingHorizontal": 30,
        "paddingVertical": 13.34,
        "width": "100%",
        "zIndex": 5000,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "flex": 1,
          },
          {},
          false,
          {
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
          },
        ]
      }
    >
      Select an option
    </Text>
    <View
      style={
        [
          {
            "marginLeft": 10,
          },
          {},
        ]
      }
    >
      <SvgMock
        fill="#0071F3"
        style={
          [
            {
              "height": 20,
              "width": 20,
            },
            {},
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`VDropDown component matches snapshot with default props 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      {
        "zIndex": 5000,
      },
      {
        "width": "40%",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onLayout={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#BED2E9",
        "borderRadius": 15,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "minHeight": 50,
        "opacity": 1,
        "paddingHorizontal": 30,
        "paddingVertical": 13.34,
        "width": "100%",
        "zIndex": 5000,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "flex": 1,
          },
          {},
          false,
          {
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
          },
        ]
      }
    >
      Select an item
    </Text>
    <View
      style={
        [
          {
            "marginLeft": 10,
          },
          {},
        ]
      }
    >
      <SvgMock
        fill="#0071F3"
        style={
          [
            {
              "height": 20,
              "width": 20,
            },
            {},
          ]
        }
      />
    </View>
  </View>
</View>
`;
